// @d11/marco/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PerformanceTracker = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const isFabricEnabled = global?.nativeFabricUIManager != null;
const isTurboModuleEnabled = global.__turboModuleProxy != null;
const PerformanceLoggerModule = isTurboModuleEnabled ? require('./NativePerformanceTracker').default : _reactNative.NativeModules.PerformanceTrackerModule;
const PerformanceTrackerView = isFabricEnabled ? require('./PerformanceTrackerViewNativeComponent').default : (0, _reactNative.requireNativeComponent)('PerformanceTrackerView');
const PerformanceTrackerViewBase = ({
  children,
  style,
  eventTimeStamp = Date.now(),
  isEnabled = true,
  meta = undefined,
  ...rest
}) => {
  let resolvedMeta;
  if (meta && isEnabled) {
    if (isFabricEnabled) {
      resolvedMeta = [];
      Object.keys(meta).forEach(key => {
        resolvedMeta.push({
          name: key,
          value: meta[key]
        });
      });
    } else {
      resolvedMeta = meta;
    }
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(PerformanceTrackerView, {
    ...rest,
    isEnabled: isEnabled,
    eventTimeStamp: eventTimeStamp,
    meta: resolvedMeta,
    style: [styles.default, style],
    children: children
  });
};
const styles = _reactNative.StyleSheet.create({
  default: {
    backgroundColor: 'rgba(255, 0, 0, 0)'
  }
});
PerformanceTrackerViewBase.displayName = 'PerformanceTracker';
PerformanceTrackerViewBase.track = (tag, time, meta) => PerformanceLoggerModule.track(tag, time, meta);
PerformanceTrackerViewBase.getLogs = () => PerformanceLoggerModule.getLogs();
PerformanceTrackerViewBase.resetLogs = options => {
  const defaultValue = {
    clearFiles: false
  };
  const finalConfig = {
    ...defaultValue,
    ...options
  };
  return PerformanceLoggerModule.resetLogs(finalConfig);
};
PerformanceTrackerViewBase.configure = config => {
  const defaultValue = {
    persistToFile: false
  };
  const finalConfig = {
    ...defaultValue,
    ...config
  };
  return PerformanceLoggerModule.configure(finalConfig);
};
const PerformanceTracker = exports.PerformanceTracker = PerformanceTrackerViewBase;
